qcengine/programs/psi4.py

```diff
@@ -53,36 +60,39 @@ def compute(self, input_model: 'ResultInput', config: 'JobConfig') -> 'Result':
         Runs Psi4 in API mode
         """
         self.found(raise_error=True)
-        import psi4
 
         # Setup the job
-        input_model = input_model.copy().dict()
-        input_model["nthreads"] = config.ncores
-        input_model["memory"] = int(config.memory * 1024 * 1024 * 1024 * 0.95)  # Memory in bytes
-        input_model["success"] = False
-        input_model["return_output"] = True
+        input_data = input_model.json_dict()
+        input_data["nthreads"] = config.ncores
+        input_data["memory"] = int(config.memory * 1024 * 1024 * 1024 * 0.95)  # Memory in bytes
+        input_data["success"] = False
+        input_data["return_output"] = True
 
-        if input_model["schema_name"] == "qcschema_input":
-            input_model["schema_name"] = "qc_schema_input"
+        if input_data["schema_name"] == "qcschema_input":
+            input_data["schema_name"] = "qc_schema_input"
 
-        scratch = config.scratch_directory
+        if config.scratch_directory:
+            input_data["scratch_location"] = config.scratch_directory
 
         if parse_version(self.get_version()) > parse_version("1.2"):
 
-            mol = psi4.core.Molecule.from_schema(input_model)
-            if (mol.multiplicity() != 1) and ("reference" not in input_model["keywords"]):
-                input_model["keywords"]["reference"] = "uhf"
+            if (input_model.molecule.molecular_multiplicity != 1) and ("reference" not in input_data["keywords"]):
+                input_data["keywords"]["reference"] = "uhf"
 
-            with scratch_directory(parent=scratch) as scrdir:
-                input_model["scratch_location"] = scrdir
-                output_data = psi4.json_wrapper.run_json(input_model)
+            # Execute the program
+            success, output = execute([which("psi4"), "--json", "data.json"], {"data.json": json.dumps(input_data)}, ["data.json"])
 
-            if "extras" not in output_data:
-                output_data["extras"] = {}
+            if success:
+                output_data = json.loads(output["outfiles"]["data.json"])
+                if "extras" not in output_data:
+                    output_data["extras"] = {}
 
-            output_data["extras"]["local_qcvars"] = output_data.pop("psi4:qcvars", None)
-            if output_data["success"] is False:
-                output_data["error"] = {"error_type": "internal_error", "error_message": output_data["error"]}
+                output_data["extras"]["local_qcvars"] = output_data.pop("psi4:qcvars", None)
+                if output_data["success"] is False:
+                    output_data["error"] = {"error_type": "internal_error", "error_message": output_data["error"]}
+            else:
+                output_data = input_data
```

Comment:
is this a general principle that upon error, output_schemus should equal input_schemus and any new info should be routed to the error field?

Thinking about CBS-flavor output_schemus where one job does not converge. Possible to return every other subjob in the output_schemus and just skip the assembly step. But the notion of returning output=input+error can be good b/c in qcf, all the subjobs will still query as run and take no time.

Reply:
In general yes. So for a failed CBS-flavor I would think we would pack all of the other jobs into `extra` still?

Reply:
ok, so principle is upon failure output_schemus == input_schemus with exceptions of success/error/extras fields. got it.